File: CliqueApp/Models/Event.swift
//
//  User.swift
//  CliqueApp
//
//  Created by Tarek Sakakini on 1/23/25.
//

import Foundation
import FirebaseCore

struct EventModel: Hashable {
    var id: String = ""
    var title: String = ""
    var location: String = ""
    var description: String = ""
    var startDateTime: Date = {
        // Initialize with current local time converted to UTC-preserving format
        let now = Date()
        let calendar = Calendar.current
        let components = calendar.dateComponents(in: TimeZone.current, from: now)
        var utcComponents = DateComponents()
        utcComponents.year = components.year
        utcComponents.month = components.month
        utcComponents.day = components.day
        utcComponents.hour = components.hour
        utcComponents.minute = components.minute
        utcComponents.second = components.second
        utcComponents.timeZone = TimeZone(identifier: "UTC")
        
        var utcCalendar = Calendar.current
        utcCalendar.timeZone = TimeZone(identifier: "UTC")!
        return utcCalendar.date(from: utcComponents) ?? Date()
    }()
    var endDateTime: Date = {
        // Initialize with current local time converted to UTC-preserving format
        let now = Date()
        let calendar = Calendar.current
        let components = calendar.dateComponents(in: TimeZone.current, from: now)
        var utcComponents = DateComponents()
        utcComponents.year = components.year
        utcComponents.month = components.month
        utcComponents.day = components.day
        utcComponents.hour = components.hour
        utcComponents.minute = components.minute
        utcComponents.second = components.second
        utcComponents.timeZone = TimeZone(identifier: "UTC")
        
        var utcCalendar = Calendar.current
        utcCalendar.timeZone = TimeZone(identifier: "UTC")!
        return utcCalendar.date(from: utcComponents) ?? Date()
    }()
    var noEndTime: Bool = false
    /// Stores user IDs (UIDs) for each RSVP bucket.
    var attendeesAccepted: [UserID] = []
    var attendeesInvited: [UserID] = []
    var attendeesDeclined: [UserID] = []
    /// Stores the hosting user's UID (legacy data may still contain emails/usernames).
    var host: UserID = ""
    var eventPic: String = ""
    var invitedPhoneNumbers: [String] = []
    var acceptedPhoneNumbers: [String] = []
    var declinedPhoneNumbers: [String] = []
    
    func initFromFirestore(eventData: [String: Any]) -> EventModel {
        var event = EventModel()
        event.id = eventData["id"] as? String ?? ""
        event.title = eventData["title"] as? String ?? "No Name"
        event.location = eventData["location"] as? String ?? "No Location"
        event.description = eventData["description"] as? String ?? ""
        event.startDateTime = (eventData["startDateTime"] as? Timestamp)?.dateValue() ?? Date()
        event.endDateTime = (eventData["endDateTime"] as? Timestamp)?.dateValue() ?? Date()
        event.noEndTime = eventData["noEndTime"] as? Bool ?? false
        event.attendeesAccepted = eventData["attendeesAccepted"] as? [String] ?? []
        event.attendeesInvited = eventData["attendeesInvited"] as? [String] ?? []
        event.attendeesDeclined = eventData["attendeesDeclined"] as? [String] ?? []
        event.host = eventData["host"] as? String ?? ""
        event.eventPic = eventData["eventPic"] as? String ?? ""
        event.invitedPhoneNumbers = eventData["invitedPhoneNumbers"] as? [String] ?? []
        event.acceptedPhoneNumbers = eventData["acceptedPhoneNumbers"] as? [String] ?? []
        event.declinedPhoneNumbers = eventData["declinedPhoneNumbers"] as? [String] ?? []
        return event
    }
}

extension EventModel {
    /// Returns the unique user IDs for all in-app participants that should be part of the chat.
    var chatParticipantUserIds: [UserID] {
        var participants = Set<UserID>()
        if !host.isEmpty {
            participants.insert(host)
        }
        attendeesAccepted.forEach { identifier in
            if !identifier.isEmpty {
                participants.insert(identifier)
            }
        }
        attendeesInvited.forEach { identifier in
            if !identifier.isEmpty {
                participants.insert(identifier)
            }
        }
        attendeesDeclined.forEach { identifier in
            if !identifier.isEmpty {
                participants.insert(identifier)
            }
        }
        return Array(participants)
    }
    
    /// Indicates whether an email should view the event from the invite tab.
    func isInviteContext(for userId: UserID) -> Bool {
        attendeesInvited.contains(userId) || attendeesDeclined.contains(userId)
    }
}

File: CliqueAndroid/app/src/main/java/com/clique/app/data/model/Event.kt
package com.clique.app.data.model

import com.clique.app.core.util.PhoneNumberFormatter
import com.google.firebase.Timestamp
import com.google.firebase.firestore.DocumentSnapshot
import java.time.Instant

private fun Timestamp?.toInstantOrNow(): Instant = this?.toDate()?.toInstant() ?: Instant.now()

data class Event(
    val id: String = "",
    val title: String = "",
    val location: String = "",
    val description: String = "",
    val startDateTime: Instant = Instant.now(),
    val endDateTime: Instant = Instant.now(),
    val noEndTime: Boolean = false,
    val attendeesAccepted: List<String> = emptyList(),
    val attendeesInvited: List<String> = emptyList(),
    val attendeesDeclined: List<String> = emptyList(),
    val host: String = "",
    val eventPic: String = "",
    val invitedPhoneNumbers: List<String> = emptyList(),
    val acceptedPhoneNumbers: List<String> = emptyList(),
    val declinedPhoneNumbers: List<String> = emptyList()
) {
    val chatParticipantUserIds: Set<String>
        get() = buildSet {
            if (host.isNotBlank()) add(host)
            attendeesAccepted.filter { it.isNotBlank() }.forEach { add(it) }
            attendeesInvited.filter { it.isNotBlank() }.forEach { add(it) }
            attendeesDeclined.filter { it.isNotBlank() }.forEach { add(it) }
        }

    fun isInviteContextFor(userId: String, phoneNumber: String): Boolean {
        val invitedById = attendeesInvited.contains(userId) || attendeesDeclined.contains(userId)
        val invitedByPhone = phoneNumber.isNotBlank() && invitedPhoneNumbers.any {
            PhoneNumberFormatter.numbersMatch(it, phoneNumber)
        }
        return invitedById || invitedByPhone
    }

    companion object {
        fun fromSnapshot(snapshot: DocumentSnapshot): Event {
            val data = snapshot.data ?: emptyMap<String, Any?>()
            return Event(
                id = data["id"] as? String ?: snapshot.id,
                title = data["title"] as? String ?: "",
                location = data["location"] as? String ?: "",
                description = data["description"] as? String ?: "",
                startDateTime = (data["startDateTime"] as? Timestamp).toInstantOrNow(),
                endDateTime = (data["endDateTime"] as? Timestamp).toInstantOrNow(),
                noEndTime = data["noEndTime"] as? Boolean ?: false,
                attendeesAccepted = data["attendeesAccepted"] as? List<String> ?: emptyList(),
                attendeesInvited = data["attendeesInvited"] as? List<String> ?: emptyList(),
                attendeesDeclined = data["attendeesDeclined"] as? List<String> ?: emptyList(),
                host = data["host"] as? String ?: "",
                eventPic = data["eventPic"] as? String ?: "",
                invitedPhoneNumbers = data["invitedPhoneNumbers"] as? List<String> ?: emptyList(),
                acceptedPhoneNumbers = data["acceptedPhoneNumbers"] as? List<String> ?: emptyList(),
                declinedPhoneNumbers = data["declinedPhoneNumbers"] as? List<String> ?: emptyList()
            )
        }
    }
}
